src/socket.rs

```diff
@@ -0,0 +1,383 @@
+pub(crate) mod listener;
+
+use std::panic;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use log::error;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
+use tokio::time;
+use tokio::time::error::Elapsed;
+use tokio::time::Instant;
+use tokio_tungstenite::tungstenite;
+use url::Url;
+
+use crate::channel::listener::{JoinedChannelReceivers, LeaveError};
+use crate::join_reference::JoinReference;
+use crate::message::*;
+use crate::socket::listener::{
+    ChannelSendCommand, ChannelSpawn, ChannelStateCommand, Connect, Join, Leave, Listener,
+    ShutdownError, StateCommand,
+};
+use crate::topic::Topic;
+use crate::{channel, Channel};
+
+const PHOENIX_SERIALIZER_VSN: &'static str = "2.0.0";
+
+/// Represents errors that occur when interacting with a [`Socket`]
+#[derive(Debug, thiserror::Error)]
+pub enum SocketError {
+    /// Occurs when the configured url is invalid for some reason
+    #[error("invalid url: {0}")]
+    InvalidUrl(Url),
+    /// Occurs when attempting to connect a client that is already connected
+    #[error("already connected")]
+    AlreadyConnected,
+    /// Occurs when attempting an operation on a disconnected client
+    #[error("not connected")]
+    NotConnected,
+    #[error("not joined to topic {0}")]
+    ChannelNotJoined(Arc<String>),
+    /// Occurs when the client fails due to a low-level connection error on the socket
+    #[error("connection error: {0}")]
+    ConnectionError(#[from] tungstenite::Error),
+}
+
+/// A [`Socket`] manages the underlying WebSocket connection used to talk to Phoenix.
+///
+/// It acts as the primary interface (along with [`Channel`]) for working with Phoenix Channels.
+///
+/// When a client is created, it is disconnected, and must be explicitly connected via [`Self::connect`].
+/// Once connected, a worker task is spawned that acts as the broker for messages being sent or
+/// received over the socket.
+///
+/// Once connected, the more useful [`Channel`] instance can be obtained via [`Self::join`]. Most functionality
+/// related to channels is exposed there.
+pub struct Socket {
+    url: Arc<Url>,
+    state_command_sender: mpsc::Sender<StateCommand>,
```

Comment:
There's a convention with channels to use tx for sender, rx for receiver

Reply:
✅